File: src/main/java/net/gtaun/shoebill/common/ColorUtils.java
package net.gtaun.shoebill.common;

import net.gtaun.shoebill.data.Color;

public class ColorUtils {
    public static Color colorBlend(Color mainColor, Color destColor, int alpha) {
        if (alpha < 0) alpha = 0;
        else if (alpha > 255) alpha = 255;

        int r = mainColor.getR();
        int g = mainColor.getG();
        int b = mainColor.getB();
        int a = mainColor.getA();

        r = r + (destColor.getR() - r) * alpha / 255;
        g = g + (destColor.getG() - g) * alpha / 255;
        b = b + (destColor.getB() - b) * alpha / 255;
        a = a + (destColor.getA() - a) * alpha / 255;

        return new Color(r, g, b, a);
    }

    public static Color hsbColor(float hue, float saturation, float brightness, int alpha) {
        return new Color((java.awt.Color.HSBtoRGB(hue, saturation, brightness) << 8) | (alpha & 0xFF));
    }

    private ColorUtils() {

    }
}

File: src/main/java/net/gtaun/shoebill/common/ColorUtils.kt
package net.gtaun.shoebill.common

import net.gtaun.shoebill.data.Color

object ColorUtils {

    @JvmStatic
    fun colorBlend(mainColor: Color, destColor: Color, alphaValue: Int): Color {
        var alpha = alphaValue
        if (alpha < 0) alpha = 0
        else if (alpha > 255) alpha = 255

        var r = mainColor.r
        var g = mainColor.g
        var b = mainColor.b
        var a = mainColor.a

        r += (destColor.r - r) * alpha / 255
        g += (destColor.g - g) * alpha / 255
        b += (destColor.b - b) * alpha / 255
        a += (destColor.a - a) * alpha / 255

        return Color(r, g, b, a)
    }

    @JvmStatic
    fun hsbColor(hue: Float, saturation: Float, brightness: Float, alpha: Int): Color =
            Color(java.awt.Color.HSBtoRGB(hue, saturation, brightness) shl 8 or (alpha and 0xFF))
}
